sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java

```diff
@@ -0,0 +1,193 @@
+    }
+    this.checkpointMark = cpMark;
+    mkTimestampPolicy(
+        Optional.ofNullable(checkpointMark).map(RabbitMqCheckpointMark::getWatermark));
+  }
+
+  private TimestampPolicy mkTimestampPolicy(Optional<Instant> prevWatermark) {
+    if (this.timestampPolicy == null) {
+      this.timestampPolicy =
+          source.spec.timestampPolicyFactory().createTimestampPolicy(prevWatermark);
+    }
+    return this.timestampPolicy;
+  }
+
+  @Override
+  public Instant getWatermark() {
+    Instant watermark = timestampPolicy.getWatermark(context, currentRecord);
+    checkpointMark.setWatermark(watermark);
+    return watermark;
+  }
+
+  @Override
+  public UnboundedSource.CheckpointMark getCheckpointMark() {
+    return checkpointMark;
+  }
+
+  @Override
+  public RabbitMqSource getCurrentSource() {
+    return source;
+  }
+
+  @Override
+  public byte[] getCurrentRecordId() {
+    return Optional.ofNullable(currentRecord).map(recordIdPolicy).get();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() {
+    return Optional.ofNullable(currentRecord).map(timestampPolicy::getTimestampForRecord).get();
+  }
+
+  @Override
+  public RabbitMqMessage getCurrent() {
+    if (currentRecord == null) {
+      throw new NoSuchElementException();
+    }
+    return currentRecord;
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    try {
+      checkpointMark.startReading();
+
+      // TODO: rework all of this
+      queueName = source.spec.queue();
+
+      ChannelLeaser.UseChannelFunction<Void> setupFn =
+          (channel) -> {
+            if (source.spec.queueDeclare()) {
+              // declare the queue (if not done by another application)
+              // channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
+              channel.queueDeclare(queueName, false, false, false, null);
+            }
+            if (source.spec.exchange() != null) {
+              if (source.spec.exchangeDeclare()) {
+                channel.exchangeDeclare(source.spec.exchange(), source.spec.exchangeType());
+              }
+              if (queueName == null) {
+                queueName = channel.queueDeclare().getQueue();
+              }
+              channel.queueBind(queueName, source.spec.exchange(), source.spec.routingKey());
+            }
+            return null;
+          };
+
+      connectionHandler.useChannel(checkpointMark.getCheckpointId(), setupFn);
+    } catch (IOException e) {
+      checkpointMark.stopReading();
+      throw e;
+    } catch (Exception e) {
+      checkpointMark.stopReading();
+      throw new IOException(e);
+    }
+    return advance();
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    try {
+      // we consume message without autoAck (we want to do the ack ourselves)
+      GetResponse delivery =
+          connectionHandler.useChannel(
+              checkpointMark.getCheckpointId(), (channel) -> channel.basicGet(queueName, false));
```

Comment:
I know this already existed, but while you're here?  💅 
```suggestion
              checkpointMark.getCheckpointId(), (channel) -> channel.basicGet(queueName, /* autoAck: */ false));
```

Reply:
also, a nit: the parens around `(channel)` should be unnecessary